### ISSUE TITLE
Employ nvrtc for run-time compilation of QUDA kernels

### ISSUE THREAD

**Issue by maddyscientist** [lattice/quda#362]
As of CUDA 7.0, run-time compilation is possible using the nvrtc library.  If we rewrite QUDA to use nvrtc, then the benefits are significant:
- Massive reduction in up-front compilation time, with all kernels only compiled on demand.  This will also have the benefit of reducing the memory resources needed for compilation.  This will significantly aid development as well as deployment, where currently tricks must be played in order to compile on systems with weak front ends.
- Improved code generation: all kernels that use nvrtc can easily use the `__launch_bounds__` qualified since the block size can be made a compile-time variable.  Experiments have shown this can improve performance of many kernels, especially double-precision ones (e.g., on a K40 DP plaquette performance goes from 300 GFLOPS to 350 GFLOPS).
- Enables compile-time autotuning: as well as the block size as mentioned above, compile-time autotuning could also include the use of texture or not, data ordering, cache intrinsics, etc.  Where CUB is deployed, CUB's policies can be swept over on a per kernel per unique parameter set.
- Reduced CPU overhead: the use of nvrtc requires the use of the CUDA driver API, which while more verbose and awkward to use, comes with the benefit of reduce CPU overhead.  This will improve strong scaling where we can be CPU bound.

The effort in deploying this is significant, though I have rough idea of how to do this.  There are multiple obstacles that have to be overcome:
- This is going to be a very intrusive change to QUDA, so to begin with we should only apply this to kernels where it will have the most impact.  That was we can refine the approach until we have something fairly clean that can be applied library wide.
- It would be nice to retain a non-nvrtc mode, which would allow QUDA to be compiled on on pre CUDA 7.0 machines.  This is not a requirement: the large Cray machines will be upgraded to CUDA 7.0, so compatibility should not be a problem.
- We need to cache the results of the compilation, so that compilation is not repeated over and over again when a kernel is returned to.  I think this would be easy to do by adding the compiled kernel pieces `CUmodule` and `CUfunction` to each instance of `std::map<TuneKey, TuneParam>`.  So when we come across a kernel that has been tuned, as well as looking up the launch parameters we look up the kernel itself.  When doing the auto tuning, for each block size we have to compile a different kernel, and only cache the best one.

The major issue that has prevented me from pursuing this prior to now has been the lack of support for compiling a templated kernel in nvrtc.  While templates within the kernel are supported, template parameters to the kernel are not supported.  Moreover, once compiled the CUDA driver API has to be able to look up the name of the kernel; if the kernel is a mangled C++ name, this can be very hard to do.  The solution here is to change the templated kernel, e.g.

``` c++
template<class T>
__global__ void testKernel(T *data)
{
  // do something
}
```

into the following

``` c++
template<class T>
__device__ void testKernel(T *data)
{
  // do something
}

extern "C" __global__ void testNew(TYPE *data) {  testKernel<TYPE>(data); }
```

and pass the compile flag `-DTYPE=type` to nvrtc to set the templated type, where `type` could be `int`, `float` etc.  While this allows us to have have effectively dynamic templates, it requires a lot of boilerplate code in the host code, since we would end up with statements like the following when compiling the kernel

``` c++
template <typename T>
void compileToPTX(const char *fname, const char *source) 
{
  const int n_flags = 1;
  char flags[n_flags][128];
  if (typeid(T) == typeid(float) {
    sprintf(flags[0], "-DTYPE=float");
  } else if (typeid(T) == typeid(int)) {
    sprintf(flags[0], "-DTYPE=int");
  }

  // compile to PTX                                                                                                                                                                                                                                                                  
  nvrtcProgram prog;
  nvrtcCreateProgram(&prog, source, fname, 0, NULL, NULL));
  nvrtcResult res = nvrtcCompileProgram(prog, n_flags, (const char**)flags);
}
```

In the above example this is manageable since we have simple types.  If, however, the argument to the kernel is some templated object then we would require a ridiculous amount of boilerplate building up the string to pass to nvrtc.

The simple work around is to employ `type_info::name` to build the string to pass to the preprocessor.  There's a gotcha here though, that being that this returns the C++ mangled name of the type and not the human-readable type that we need to be able to compile.  Fortunately, we can employ the `cxxabi` which provides a portable method to accessing the demangled name.

``` c++
#include <cxxabi.h> // for compiling with clang on Ubuntu, the libc++abi-dev package is required
#include <string>
#include <typeinfo>
std::string demangle(const char* mangled)
{
  using std;
  int status;
  unique_ptr<char[], void (*)(void*)> result(abi::__cxa_demangle(mangled, 0, 0, &status), free);
  return result.get() ? string(result.get()) : "error occurred";
}
```

This makes it very easy to build to compile templates on demand, and this should work for arbitrary templated types.

``` c++
template <typename T>
void compileToPTX(const char *fname, const char *source)
{
  const int n_flags = 1;
  char flags[n_flags][128];
  sprintf(flags[0], "-DTYPE=%s", demangle(typeid(T).name()).c_str()); // magic is here!

  // compile to PTX                                                                                                                                                                                                                                                                 
  nvrtcProgram prog;
  nvrtcCreateProgram(&prog, source, fname, 0, NULL, NULL));
  nvrtcResult res = nvrtcCompileProgram(prog, n_flags, (const char**)flags);
}
```

**Comment by mikeaclark** [comment]
After some experimentation this looks like it will be a pain to implement.  The main issue surrounds the many headers that QUDA kernels typically include, these cannot include any non-GPU code.

The good news is that the changes required to do this, will increase the modularity of QUDA and help with the ongoing maintainability.  I suspect that having cmake working will help also in this regard, as it should ease this increased modularity approach.

An additional motivation for this: when the multigrid branch is built, with all options enabled, the binary is approaching 1GB.  This is rather ridiculous.

On the implementation, what I favour at the moment is extending the `Tunable` class with a `TunableJIT` class.  This class will incorporate new member functions for compiling the kernels to ptx and then to the binary code.  With this implemented, we can easily retain the old-style `Tunable` launcher and port each kernel individually.

**Comment by mathiaswagner** [comment]
Just adding here as a possible optimization to explore with RTC:
It might not be straightforward but we could explore templating e.g. Dslash-Kernels also with lattice dimensions. If the compiler is smart enough that might save us some of the integer operations for indexing ...

**Comment by mathiaswagner** [comment]
Just to think about: What will we do with tuning of launch parameters? Add a driver-dependency to the tunecache?

**Comment by maddyscientist** [comment]
Probably no need to template on lattice dimensions.  The primary benefit from this is it allows integer division by constants to be done instead of using variables.  Division by constants can be done using magic numbers, however, the same effect can be achieved using @milakov's fast integer division class
(https://github.com/milakov/int_fastdiv).

**Comment by maddyscientist** [comment]
Yes, we will need to record the driver version in the tunecache, but this is a minor detail since the driver will change far less frequently than say the git tag.

**Comment by mathiaswagner** [comment]
Sure. It might save some registers as well but fastdiv might be the same.
I know the driver is not likely to change frequently it just came to my mind.

**Comment by maddyscientist** [comment]
This issue will (finally) be closed by pull #725
